unpublished/build-system-abstraction-njs.rst

```diff
@@ -0,0 +1,708 @@
+combination of wheels and sdists. In a command like ``pip install
+lxml==2.4.0``, pip is acting as an integration frontend.
+
+
+==============
+ Source trees
+==============
+
+We retroactively declare the legacy source tree format involving
+``setup.py`` to be "version 0". We don't try to specify it further;
+its de facto specification is encoded in the source code and
+documentation of ``distutils``, ``setuptools``, ``pip``, and other
+tools.
+
+A "version 1" (or greater) source tree is any directory which contains
+a file named ``pypackage.json``. (If a tree contains both
+``pypackage.json`` and ``setup.py`` then it is a version 1+ source
+tree and the ``setup.py`` is ignored; this allows packages to include
+a ``setup.py`` for compatibility with old build frontends, while using
+the new system with new build frontends.)
+
+This file has the following schema. Extra keys are ignored.
+
+schema
+    The version of the schema. This PEP defines version "1".  Defaults to "1"
+    when absent. All tools reading the file MUST error on an unrecognised
+    schema version.
+
+bootstrap_requires
+    Optional list of PEP 508 dependency specifications that the
+    build frontend must ensure are available before invoking the build
+    backend. For instance, if using flit, then the requirements might
+    be::
+
+        "bootstrap_requires": ["flit"]
+
+build_backend
+    A mandatory string naming a Python object that will be used to
+    perform the build (see below for details). This is formatted
+    following the same ``module:object`` syntax as a ``setuptools``
+    entry point. For instance, if using flit, then the build system
+    might be specified as::
+
+        "build_system": "flit.api:main"
+
+    and this object would be looked up by executing the equivalent of::
+
+        import flit.api
+        backend = flit.api.main
+
+    It's also legal to leave out the ``:object`` part, e.g. ::
+
+        "build_system": "flit.api"
+
+    which acts like::
+
+        import flit.api
+        backend = flit.api
+
+    Formally, the string should satisfy this grammar::
+
+        identifier = (letter | '_') (letter | '_' | digit)*
+        module_path = identifier ('.' identifier)*
+        object_path = identifier ('.' identifier)*
+        entry_point = module_path (':' object_path)?
+
+    And we import ``module_path`` and then lookup
+    ``module_path.object_path`` (or just ``module_path`` if
+    ``object_path`` is missing).
+
+
+=========================
+ Build backend interface
+=========================
+
+The build backend object is expected to have attributes which provide
+some or all of the following hooks. The common ``config_settings``
+argument is described after the individual hooks::
+
+  def get_build_requires(config_settings):
+      ...
+
+This hook MUST return an additional list of strings containing PEP 508
+dependency specifications, above and beyond those specified in the
+``pypackage.json`` file. Example::
```

Comment:
Yes, it might (and in some cases surely will) depend on OS, python version, the `config_settings`, etc. That's why it's a function rather than a static list :-)